src/backend/Functions/ProcessWorkflowQueue.cs

```diff
@@ -14,17 +15,43 @@ public static class ProcessWorkflowQueue
         [FunctionName("ProcessWorkflowQueue")]
         public static async Task Run([QueueTrigger("apiportworkflowqueue")]WorkflowQueueMessage workflowMessage,
             [Queue("apiportworkflowqueue")]ICollector<WorkflowQueueMessage> workflowMessageQueue,
-            ILogger log)
+            ILogger log,
+            CancellationToken cancelToken)
         {
-            log.LogInformation("Processing message {SubmissionId}, stage {Stage}", workflowMessage.SubmissionId, workflowMessage.Stage);
+            Action<WorkflowQueueMessage> cancelAsync = (msg) =>
+            {
+                log.LogError("Timeout during workflow action execution.SubmissionId: { msg.SubmissionId}, Stage: { msg.Stage}", workflowMessage.SubmissionId, workflowMessage.Stage);
+                throw new TimeoutException($"Timeout during workflow action execution. SubmissionId: {msg.SubmissionId} Stage: {msg.Stage}.");
+            };
+
+            //This cancellation token will let us know if this function has timed out.  When it times out, throw an exception so that the 
+            //function completes with a failure.  This will ensure that the message is removed from apiportworkflowqueue and automatically added to apiportworkflowqueue-poison so that
+            //we can treat a function timeout as an ordinary failure where the end user must resubmit their submission for analysis.
```

Comment:
Instead of the `if/else` you can call `cancelToken.ThrowIfCancelRequested()` and it will throw if needed.

Reply:
I didn't do this because ThrowIfCancellation doesn't allow me to pass my own message; I wanted the exception to include the submission id and stage for being able to query easily in AI.  Also, I want to ensure LogError can be called.